Forbid unsafe types in records

333fred opened the pull request:
[The specification](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/records.md#members-of-a-record-type) for records forbids the use of unsafe types as an instance field of the type. However, we didn't check that in all cases; we only checked when generating the `Equals` method for the record type. This meant that if the user provided their own definition of `Equals`, they could bypass this restriction. We also did not check for nested unsafe types, so `int*[]` was permitted in any scenario. We have 2 options for fixing this:

1. Treat it as a specification bug, and update the spec to follow the current behavior. We'd therefore fix https://github.com/dotnet/roslyn/issues/66312 by moving the diagnostic to SourceNamedTypeSymbol.AfterMembersChecks, and only error when the user didn't provide their own implementation of `Equals`.
2. Treat it as a compiler bug, and forbid use of unsafe types in all scenarios.

I've opted for path 2 here, but we can discuss whether this is too aggressive and if we should opt for path 1 instead. Fixes https://github.com/dotnet/roslyn/issues/66312.

333fred commented:
We also need to decide what to do with manually-implemented properties that have unsafe types. As pointed out by @hamarb123, even with this change we'll still generate invalid IL for `PrintMembers` in such a case, where we cast the pointer to `object`. By my reading of the spec, it says we should make it work; nothing about the definition of `PrintMembers` specifies that this cast needs to occur, and we could instead convert the pointer to `IntPtr` before doing the print. The main question is whether we _should_ do this, or just forbid all pointer types in `records`, period.

AlekseyTs commented on docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md:
```diff
@@ -156,3 +156,30 @@ struct S
+
+***Introduced in Visual Studio 2022 version 17.14***
+
+The specification for `record class` and `record struct` types indicated that any unsafe types are disallowed. However, this was not
```
>indicated that any unsafe types are disallowed [](http://example.com/codeflow?start=63&length=46)

This sounds too broad. The spec says:

> It is an error for an instance field of a record to have an unsafe type.

AlekseyTs commented on docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md:
```diff
@@ -156,3 +156,30 @@ struct S
+enforced correctly in 2 scenarios:
+
+1. When the `record class` or `record struct` type defined its own `Equals` implementation.
+2. When the field type only used the unsafe type in a nested context, such as `int*[]`.
```
>When the field type only used the unsafe type in a nested context, such as `int*[]`. [](http://example.com/codeflow?start=3&length=84)

I couldn't find clear definition of "unsafe type" in the language spec. Therefore, it is not clear whether such array type should be considered as unsafe.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberFieldSymbol.cs:
```diff
@@ -57,15 +57,20 @@ protected void TypeChecks(TypeSymbol type, BindingDiagnosticBag diagnostics)
-                diagnostics.Add(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, TypeSyntax?.Location ?? this.GetFirstLocation(), type);
+                diagnostics.Add(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, getTypeErrorLocation(), type);
+            }
+            else if (!this.IsStatic && type.ContainsPointerOrFunctionPointer() && (ContainingType.IsRecord || ContainingType.IsRecordStruct))
```
>(ContainingType.IsRecord || ContainingType.IsRecordStruct) [](http://example.com/codeflow?start=82&length=58)

I assume this check is cheaper than the type check for pointers. Consider performing reordering the checks.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs:
```diff
@@ -1839,9 +1839,18 @@ protected virtual void ValidatePropertyType(BindingDiagnosticBag diagnostics)
+            if (this.IsAutoPropertyOrUsesFieldKeyword)
             {
-                diagnostics.Add(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, TypeLocation, type);
+                if (!this.IsStatic && type.ContainsPointerOrFunctionPointer() && (ContainingType.IsRecord || ContainingType.IsRecordStruct))
```
>(ContainingType.IsRecord || ContainingType.IsRecordStruct) [](http://example.com/codeflow?start=81&length=58)

Similar suggestion

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:
```diff
@@ -1707,9 +1802,8 @@ public unsafe record C
-            comp.VerifyEmitDiagnostics(
-                // (12,22): error CS0723: Cannot declare a variable of static type 'C2'
+            var comp = CreateCompilation(new[] { src, IsExternalInitTypeDefinition }, options: TestOptions.UnsafeDebugDll, targetFramework: TargetFramework.Mscorlib461);
+            DiagnosticDescription[] expected = [// (12,22): error CS0723: Cannot declare a variable of static type 'C2'
```
>// (12,22): error CS0723: Cannot declare a variable of static type 'C2' [](http://example.com/codeflow?start=48&length=71)

Consider placing the comment on the next line

AlekseyTs commented:
It looks like more test base-lines need an adjustment

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed your feedback. The spec part of this change is at https://github.com/dotnet/csharplang/pull/9050, clarifying what "unsafe types" is actually supposed to mean.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:
```diff
@@ -136,12 +136,7 @@ internal override void GenerateMethodBody(TypeCompilationState compilationState,
-                            foundBadField = true;
-                        }
-                        else if (parameterType.IsRestrictedType())
+                        if (parameterType.ContainsPointerOrFunctionPointer() || parameterType.IsRestrictedType())
```
>ContainsPointerOrFunctionPointer [](http://example.com/codeflow?start=42&length=32)

I thought we decided to allow arrays of pointers

333fred replied:
Thanks for pointing it out, I added a test that would have failed.

AlekseyTs commented:
Done with review pass (commit 3)

AlekseyTs reviewed (approved):
LGTM (commit 4)

333fred commented:
@dotnet/roslyn-compiler for a second review

hamarb123 commented:
Did you end up blocking manually implemented unsafe properties, or allowing them, out of curiosity. Based on the spec change I assume they're allowed?

333fred commented:
> Did you end up blocking manually implemented pointer type properties, or allowing them (and fixing the `PrintMembers`), out of curiosity. Based on the spec change I assume they're allowed?

No change was made to locations without fields.

333fred commented:
@dotnet/roslyn-compiler for a second review

cston commented on docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md:
```diff
@@ -193,3 +193,23 @@ namespace Microsoft.CodeAnalysis;
+
+## `record` and `record struct` types cannot define pointer type members, even when providing their own Equals implementations
+
+PROTOTYPE: Figure what version this is going into
```
Remove?

cston commented on docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md:
````diff
@@ -193,3 +193,23 @@ namespace Microsoft.CodeAnalysis;
+The compiler now correctly forbids this.
+
+```cs
+unsafe record struct R1(
````
>R1 [](http://example.com/codeflow?start=21&length=2)

`R`, to match the `Equals(R)` declaration.

cston commented on src/Compilers/CSharp/Test/Emit3/Semantics/RecordTests.cs:
```diff
@@ -1271,12 +1272,73 @@ unsafe record C(
+        [Fact]
+        [WorkItem(48115, "https://github.com/dotnet/roslyn/issues/48115")]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/66312")]
+        public void RestrictedTypesAndPointerTypes_WithEquals()
```
>RestrictedTypesAndPointerTypes_WithEquals [](http://example.com/codeflow?start=20&length=41)

Since the expected behavior is identical to the previous test, consider using `[Theory]` with `bool withEquals` parameter.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:
```diff
@@ -1526,15 +1527,71 @@ unsafe record struct C(
+        [Fact]
+        [WorkItem(48115, "https://github.com/dotnet/roslyn/issues/48115")]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/66312")]
+        public void RestrictedTypesAndPointerTypes_WithEquals()
```
>RestrictedTypesAndPointerTypes_WithEquals [](http://example.com/codeflow?start=20&length=41)

Consider combining with previous test with `[Theory]` and `bool withEquals` parameter, so it's clear we expect the behavior to be the same.

cston commented on src/Compilers/CSharp/Test/Emit3/Semantics/RecordTests.cs:
```diff
@@ -1321,12 +1383,14 @@ public unsafe record C
+        [Fact]
         [WorkItem(48115, "https://github.com/dotnet/roslyn/issues/48115")]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/66312")]
         public void RestrictedTypesAndPointerTypes_NominalMembers_AutoProperties()
```
>RestrictedTypesAndPointerTypes_NominalMembers_AutoProperties [](http://example.com/codeflow?start=20&length=60)

Consider testing at least one case with `field`.
```csharp
unsafe record C
{
    int* P { get => field; }
}
```

cston commented on docs/compilers/CSharp/Compiler Breaking Changes - DotNet 10.md:
```diff
@@ -196,7 +196,7 @@ public class EmbeddedAttribute : Attribute {}
 ## `record` and `record struct` types cannot define pointer type members, even when providing their own Equals implementations
 
-PROTOTYPE: Figure what version this is going into
+***Introduced in Visual Studio 2022 version 17.14***
```
Duplicated below.
